esp32c2-hal/examples/interrupt_preemption.rs

```diff
@@ -0,0 +1,147 @@
+//! Software Interrupts
+//!
+//! An example of how software interrupts can be raised and reset
+//! Should rotate through all of the available interrupts printing their number
+//! when raised.
+
+#![no_std]
+#![no_main]
+
+use core::cell::RefCell;
+
+use critical_section::Mutex;
+use esp32c2_hal::{
+    clock::ClockControl,
+    interrupt::{self},
+    peripherals::{self, Peripherals},
+    prelude::*,
+    riscv,
+    system::{SoftwareInterrupt, SoftwareInterruptControl},
+    timer::TimerGroup,
+    Rtc,
+};
+use esp_backtrace as _;
+
+static SWINT: Mutex<RefCell<Option<SoftwareInterruptControl>>> = Mutex::new(RefCell::new(None));
+
+#[entry]
+fn main() -> ! {
+    let peripherals = Peripherals::take();
+    let mut system = peripherals.SYSTEM.split();
+    let clockctrl = system.clock_control;
+    let sw_int = system.software_interrupt_control;
+    let clocks = ClockControl::boot_defaults(clockctrl).freeze();
+
+    let mut rtc = Rtc::new(peripherals.RTC_CNTL);
+    let timer_group0 = TimerGroup::new(
+        peripherals.TIMG0,
+        &clocks,
+        &mut system.peripheral_clock_control,
+    );
+    let mut wdt0 = timer_group0.wdt;
+
+    // Disable watchdog timers
+    rtc.swd.disable();
+    rtc.rwdt.disable();
+    wdt0.disable();
+
+    critical_section::with(|cs| SWINT.borrow_ref_mut(cs).replace(sw_int));
+
+    interrupt::enable(
+        peripherals::Interrupt::FROM_CPU_INTR0,
+        interrupt::Priority::Priority1,
+    )
+    .unwrap();
+    interrupt::enable(
+        peripherals::Interrupt::FROM_CPU_INTR1,
+        interrupt::Priority::Priority2,
+    )
+    .unwrap();
+    interrupt::enable(
+        peripherals::Interrupt::FROM_CPU_INTR2,
+        interrupt::Priority::Priority2,
+    )
+    .unwrap();
+    interrupt::enable(
+        peripherals::Interrupt::FROM_CPU_INTR3,
+        interrupt::Priority::Priority4,
```

Comment:
I'm noticing the example doesn't showcase the behavior of max-priority interrupts; would it be worth bumping this to priority 15, maybe?

Reply:
I've bumped this to max prio now because why not, but the stack smashing behavior you mentioned wouldn't occur without raising an interrupt during the context restore, which isn't really achievable with software interrupts only as far as i see. Either way, i think we should be fine on this front anyway.